src/Azure/Shared/CosmosDB/AzureCosmosDBConnectionFactory.cs

```diff
@@ -0,0 +1,51 @@
+using System.Text.Json;
+using Microsoft.Extensions.DependencyInjection;
+
+#if ORLEANS_CLUSTERING
+namespace Orleans.Clustering.CosmosDB;
+#elif ORLEANS_PERSISTENCE
+namespace Orleans.Persistence.CosmosDB;
+#elif ORLEANS_REMINDERS
+namespace Orleans.Reminders.CosmosDB;
+#elif ORLEANS_STREAMING
+namespace Orleans.Streaming.CosmosDB;
+#elif ORLEANS_DIRECTORY
+namespace Orleans.GrainDirectory.CosmosDB;
+#else
+// No default namespace intentionally to cause compile errors if something is not defined
+#endif
+
+internal class AzureCosmosDBConnectionFactory
+{
+    public static ValueTask<CosmosClient> CreateCosmosClient(IServiceProvider serviceProvider, AzureCosmosDBOptions options)
+    {
+        if (options.CosmosDBClientFactory is not null)
+        {
+            return options.CosmosDBClientFactory(serviceProvider);
+        }
+
+        var clusterOptions = serviceProvider.GetRequiredService<IOptions<ClusterOptions>>().Value;
+
+        var jsonSerializerOptions = options.SerializerOptions ?? new JsonSerializerOptions
```

Comment:
Maybe we should add the converters that are added in `OrleansJsonSerializer.GetDefaultSerializerSettings`?

Reply:
To be honest, I'd like to at some point get rid of the reflection-based JSON serialization across all the projects. System.Text.Json has a source generator which we could leverage for all our internal types. Let's do it on a separated PR.